presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,552 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * implementation that uses groupId in the hash key, so that we may store all groupId x value -> count in one giant map. The value, however,
+ * is normalized into a single ValueBuilder. Further, we do not have per-group instances of objects. In order to construct a histogram for a group on
+ * demand, a linked list for each group is stored that overlays all virtual nodes (a shared index, i, across parallel arrays).
+ * <p>
+ * <pre>
+ * eg,
+ * heads[100] -> 3
+ *
+ * means the first piece of data is (values[valuePositions[3]], counts[3])
+ * If next[3] is 10, then we look at (values[valuePositions[10]], counts[10])
+ *
+ * and so on to construct the value, count pairs needed. An iterator-style function, readAllValues, exists and the caller may do whatever want with the
+ * values.
+ *
+ * </pre>
+ */
+
+public class GroupedTypedHistogram
+        implements TypedHistogram
+{
+    static final float MAX_FILL_RATIO = 0.5f;
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final int EMPTY_BUCKET = -1;
+    private static final int NULL = -1;
+    private static final ProbeType PROBE_TYPE = ProbeType.SUM_OF_SQUARE;
+    private final int bucketId;
+
+    private final Type type;
+    private final BlockBuilder values;
+    private final BucketNodeFactory bucketNodeFactory;
+    // these parallel arrays represent a node in the hash table
+    private final LongBigArray counts;
+    // need to store the value of a node's groupId
+    private final LongBigArray groupIds;
+    // array of nodePointers (index in counts, valuePositions)
+    private final IntBigArray nextPointers;
+    // per groupId, we have a pointer to the first in the list of nodes for this group
+    private final LongBigArray headPointers;
+
+    private IntBigArray buckets;
+    private int nextNodePointer;
+    private int mask;
+    private int bucketCount;
+    private int maxFill;
+    // at most one thread uses this object at one time, so this must be set to the group being operated on
+    private long currentGroupId = -1;
+    private long numberOfGroups = 1;
+    private final LongBigArray valueRawHashList = new LongBigArray();
+    private final ValueStore valueStore;
+    private final IntBigArray valuePositions;
+
+    public GroupedTypedHistogram(
+            Type type,
+            int expectedCount,
+            BlockBuilder values)
+    {
+        this.type = type;
+        this.bucketId = expectedCount;
+        this.bucketCount = computeBucketCount(expectedCount, MAX_FILL_RATIO);
+        this.mask = bucketCount - 1;
+        this.maxFill = calculateMaxFill(bucketCount, MAX_FILL_RATIO);
+        this.values = values;
+
+        checkArgument(expectedCount > 0, "expectedSize must be greater than zero");
+
+        buckets = new IntBigArray(-1);
+        buckets.ensureCapacity(bucketCount);
```

Comment:
Call `resizeNodeArrays()` after all are initialized instead of all these calls to `ensureCapacity`

Reply:
i'm not sure I see what you are referring to? resizeNodeArrays() is called in the add()->rehash() code path.  

are you saying i should *ADD* a call here? to make sure they are properly sized?

i think they are manually sized above that location, but are you saying ensureCapacity should be called first?  might need to talk to you on msgr or in person to understand this

Reply:
done